### ISSUE TITLE
[Java] NP_GUARANTEED_DEREF false negative

### ISSUE THREAD

**Issue by RuS2m** [spotbugs/spotbugs#2693]
Hi there,

I found a false negative for the rule [NP_GUARANTEED_DEREF](https://spotbugs.readthedocs.io/en/latest/bugDescriptions.html#np-null-value-is-guaranteed-to-be-dereferenced-np-guaranteed-deref)/[NP_ALWAYS_NULL](https://spotbugs.readthedocs.io/en/latest/bugDescriptions.html#np-null-pointer-dereference-np-always-null) while using Spotbugs

Seems like some checks, at least null dereference ones, do not apply inside exception catch clauses.

The minimal case below:

```java
String str = null;
try {
    throw new NullPointerException(); // Could be any exception here
} catch (Exception e) {
    str.length() // this is throwing NPE
}
```

Versions:
- the issue was reproduced on Spotbugs versions `4.4.1`, `4.5.1`, `4.8.1` and on the mainline up to [this commit](https://github.com/spotbugs/spotbugs/commit/00bd670861d465a52e24a607e249383dd267a155) on spotbugsTestCases
- Java version: `java 17.0.9-oracle`
- Gradle version: `gradle 8.4`

**Comment by ThrawnCA** [comment]
I don't think it's feasible to analyse the content of `try` blocks from the perspective of the `catch` blocks, since you explicitly don't know how much of the try block was executed. Sure, in this case a human can look at it and say that nothing ever tries to populate `str`, but as a general solution? I have doubts.

**Comment by RuS2m** [comment]
Thanks for your prompt reply, @ThrawnCA!

Definitely understand that it's hard to analyze the content of `try` blocks since they could be partially executed.

However in the case I described in the ticket, is there any possibility to detect that the variable `str` is not updated in the `try` block and therefore if the exception will be thrown, the catch block fail with NPE?

**Comment by RuS2m** [comment]
Any updates on this, @ThrawnCA? Is the suggestion I proposed in my comment above impossible to implement?
I would be happy to contribute to the project myself with this change, but I can't find the internal documentation for the project

**Comment by ThrawnCA** [comment]
> Any updates on this, @ThrawnCA? Is the suggestion I proposed in my comment above impossible to implement?

Sorry, I help with code reviews, but the intricacies of writing a detector are beyond me.

> I would be happy to contribute to the project myself with this change, but I can't find the internal documentation for the project

What kind of documentation are you after? The file containing this detector is https://github.com/spotbugs/spotbugs/blob/master/spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindNullDerefsInvolvingNonShortCircuitEvaluation.java
